# 01. Importing Libraries

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

# 02. Importing Data

In [2]:
# Project folder pathway
path = r'C:\Users\SD.DESKTOP-45BHU58\OneDrive\Desktop\CF\Section 4 - Python\Instacart Basket Analysis'

In [3]:
path

'C:\\Users\\SD.DESKTOP-45BHU58\\OneDrive\\Desktop\\CF\\Section 4 - Python\\Instacart Basket Analysis'

In [4]:
# Importing products.csv from Original Data

df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'))

### 1. If you haven’t performed the consistency checks covered in this Exercise on your df_prods dataframe, do so now

### Mixed-Type Data

In [24]:
for col in df_prods.columns.tolist():
  weird = (df_prods[[col]].applymap(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods[weird]) > 0:
    print (col)

product_name


In [25]:
# Changing 'product_name' to string

df_prods['product_name'].astype('string')

0                               Chocolate Sandwich Cookies
1                                         All-Seasons Salt
2                     Robust Golden Unsweetened Oolong Tea
3        Smart Ones Classic Favorites Mini Rigatoni Wit...
4                                Green Chile Anytime Sauce
                               ...                        
49688            Vodka, Triple Distilled, Twist of Vanilla
49689                   En Croute Roast Hazelnut Cranberry
49690                                     Artisan Baguette
49691           Smartblend Healthy Metabolism Dry Cat Food
49692                               Fresh Foaming Cleanser
Name: product_name, Length: 49693, dtype: string

In [28]:
# Checking data types

df_prods.dtypes

product_id         int64
product_name      object
aisle_id           int64
department_id      int64
prices           float64
dtype: object

### Missing Values

In [29]:
# Find missing observations

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [30]:
# Create df with just missing product_name values

df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [31]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [32]:
# Can't impute values since product_name values are string

In [32]:
# Number of rows in df_prods

df_prods.shape

(49693, 5)

In [33]:
# Create new df with non-missing values

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [34]:
df_prods_clean.shape

(49677, 5)

In [36]:
# Double checking missing values

df_prods_clean.isnull().sum()

product_id       0
product_name     0
aisle_id         0
department_id    0
prices           0
dtype: int64

### Duplicates

In [37]:
# Create new df with duplicate records

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [38]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [39]:
# Checking row numbers in df_prods_clean df

df_prods_clean.shape

(49677, 5)

In [20]:
# Creating new df w/ no dups

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [21]:
df_prods_clean_no_dups.shape

(49672, 5)

### 02. Run the df.describe() function on your df_prods dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further

In [42]:
# Cleaning df_prods df name

df_prods_checked = df_prods_clean_no_dups

In [68]:
df_prods_checked.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

The count and max for product_id do not match. The max price of $99,999 also looks odd, no grocery store item would be this high.

### 03. Check for mixed-type data in your df_ords dataframe

In [48]:
# Importing wrangled orders.csv from Prepared Data

df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'))

In [52]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col + ' has mixed type data')
  else:
    print (col + ' data type is uniform')     

Unnamed: 0 data type is uniform
order_id data type is uniform
user_id data type is uniform
order_number data type is uniform
orders_day_of_the_week data type is uniform
order_hour_of_day data type is uniform
days_since_prior_order data type is uniform


No column contains mixed type data

### 04. If you find mixed-type data, fix it. The column in question should contain observations of a single data type

N/A

### 05. Run a check for missing values in your df_ords dataframe. In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [53]:
# Find missing observations

df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_the_week         0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

The 'days_since_prior_order' column contains 206,209 missing records. These are most likely coming from customers who are ordering from the first time, hence there is no prior date.

### 06. Address the missing values using an appropriate method. In a markdown cell, explain why you used your method of choice

The 206,209 missing records are coming from first-time customers. This information should not be deleted since it is important in letting us know which customers are new. A new variable needs to be created to act like a flag based on the missing value. This new variable should be named 'first_order' to indicate if a customer is placing an order for the first time, which will also help explain the missing records from 'days_since_prior_order'.

In [55]:
# Creating df with new orders

df_ords_clean = df_ords

In [58]:
# 'first_order' column will have True values for missing records, otherwise it will be False

df_ords_clean['first_order'] = df_ords['days_since_prior_order'].isnull() == True

In [60]:
df_ords_clean.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_the_week  \
0           0   2539329        1             1                       2   
1           1   2398795        1             2                       3   
2           2    473747        1             3                       3   
3           3   2254736        1             4                       4   
4           4    431534        1             5                       4   

   order_hour_of_day  days_since_prior_order  first_order  
0                  8                     NaN         True  
1                  7                    15.0        False  
2                 12                    21.0        False  
3                  7                    29.0        False  
4                 15                    28.0        False

### 07. Run a check for duplicate values in your df_ords data. In a markdown cell, report your findings and propose an explanation for any duplicate values you find

In [62]:
# Creating new df with duplicate records

df_ords_clean_dups = df_ords_clean[df_ords_clean.duplicated()]

In [63]:
df_ords_clean_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_the_week, order_hour_of_day, days_since_prior_order, first_order]
Index: []

### 08. Address the duplicates using an appropriate method. In a markdown cell, explain why you used your method of choice

N/A

### 09. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names

In [65]:
# Exporting clean df_ords df

df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

In [67]:
# Exporting clean df_prods df

df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))